[28][29][31] Width and Height block settings

PaulREnglish opened the pull request:
## Description

Fixes https://github.com/bigbite/image-comparison/issues/28, https://github.com/bigbite/image-comparison/issues/29 & https://github.com/bigbite/image-comparison/issues/31

Assigns the Image Comparisons `containerHeight` and `containerWidth` values a default of `null`. This is to avoid an issue where the block won't scale to the new width if the `settings.layout.contentSize` setting in `theme.json` is changed and the user has not set a width yet.

Fixes an issue where the alignment options in the block toolbar have no effect due to it being overridden by the `containerHeight` and `containerWidth` attribute styles that are passed into the `ResizableBox` component.  Instead, we now set the `containerWidth` and `containerHeight` to `auto` if the alignment attribute has been set.


## Change Log

- Fixes an issue where `containerHeight` and `containerWidth` values not specifically set the by user would not scale when the `settings.layout.contentSize` setting in theme.json would change.
- Fixes an issue where the alignment options in the block toolbar have no effect.

## Steps to test

### Issue https://github.com/bigbite/image-comparison/issues/28

1. Add an Image Comparison block
2. Add both images
3. Save
4. Increase the settings.layout.contentSize value in theme.json
5. Refresh page
6. See that the Image Comparison block is now displaying correctly in the editor and on the frontend


### Issue https://github.com/bigbite/image-comparison/issues/29

1. Add an Image Comparison block
2. Add both images
3. Apply a settings.layout alignment value
4. The width and height settings within the block sidebar should disappear.
5. The alignment of the block should be reflected correctly both within the editor and on the frontend.

## Screenshots/Videos

http://bigbite.im/v/KfN5NC (https://github.com/bigbite/image-comparison/issues/28)

http://bigbite.im/v/3piqUx (https://github.com/bigbite/image-comparison/issues/29)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

PaulREnglish committed:
fix: assign height and width a default value of null

Fixes an issue where editing settings.layout.contentSize in theme.json would not change the width of the block.

PaulREnglish committed:
fix(image-comparison): move block props into own function

This was needed to fix 'Arrow function has a complexity of 12. Maximum allowed is 10' JS lint error

PaulREnglish committed:
Revert "fix(image-comparison): move block props into own function"

This reverts commit 8c657d88c6ee30cc663709eb3558ab1023476715.

PaulREnglish committed:
fix(image comparison): create function for getting container size

This was needed to fix 'Arrow function has a complexity of 12. Maximum allowed is 10' JS lint error

PaulREnglish committed:
docs: fix invalid getContainerSize typehinting in docblock

PaulREnglish committed:
fix(image-comparison): allows alignment settings to affect the block

PaulREnglish committed:
docs: correct return param type in getContainerSize docblock

PaulREnglish committed:
Merge branch 'fix/width-height-default-null' into fix/no-layout-alignment-override

PaulREnglish committed:
fix: fix alignment functionality on front end

PaulREnglish committed:
fix: hide width/height settings when alignment is set

PaulREnglish committed:
docs: improve getContainerSize docblock description

PaulREnglish committed:
Merge branch 'fix/width-height-default-null' into fix/no-layout-alignment-override

chrishbite commented on src/blocks/image-comparison/components/Edit.js:
```diff
@@ -46,10 +45,46 @@ const Edit = ({ attributes, setAttributes, clientId }) => {
-    containerWidth,
   } = attributes;
 
+  /**
+   * Get the container's size. If this has not been set by
+   * the user then we overwrite the default size of the block
+   * with the theme's defined contentSize, if it exists.
+   *
+   * @returns {
+   *  width: string
+   *  height: string
+   * } The size containing the height and width of the block's container.
+   */
+  const getContainerSize = () => {
+    if (attributes.align) {
+      return {
+        containerWidth: 'auto',
+        containerHeight: 'auto',
+      };
+    }
+
+    let containerHeight = '500px';
+    if (attributes.containerHeight) {
+      containerHeight = attributes.containerHeight;
+    }
+
+    let containerWidth = '500px';
+    if (attributes.containerWidth) {
+      containerWidth = attributes.containerWidth;
+    } else if (contentWidth) {
+      containerWidth = contentWidth;
+    }
+
+    return {
+      containerHeight,
+      containerWidth,
+    };
+  };
+
+  const { containerHeight, containerWidth } = getContainerSize();
```
I don't think this function will be needed? If the `containerHeight` and `containerWidth` attributes aren't set then the block just needs to respond to it's parent element, the nullish coalescing operator (`??`) might hopefully be enough to handle this?

I'm not sure where the `500px` default width and height values have come from as currently this will lead to a visual discrepancy between editor and frontend, steps to reproduce this;

- Add Image Comparison block
- Add 2 images
- Save

The block in editor will end up with the height of `500px`, if you then visit the frontend you'll see that the block is displaying as expected (height scaling with the width), are the `500px` values required?

chrishbite replied:
Hey @PaulREnglish thanks for the explanation, we should try and avoid setting any width or height pixel values until the user specifically sets them and let the image respond to it's container by default.

I'm not sure of the reason they were set to default to `500px` originally as well, I may be overlooking something here but please let me know if I can assist with anything on this 👍

PaulREnglish replied:
Will do thanks!

chrishbite reviewed (changes requested):
Hey @PaulREnglish  thanks for the work on this, I've left a couple of changes below on this please.

I've also consolidated the dependent PRs description/testing steps into this one and closed [38](https://github.com/bigbite/image-comparison/pull/38) as these will be easier to review and test them together with code overlapping and being dependent on each other.


**1)** These changes will generate the below PHP warnings from `image-comparison/render.php` with default usage.

```
Warning: Undefined array key "align"
Warning: Undefined array key "containerHeight"
Warning: Undefined array key "align"
Warning: Undefined array key "containerWidth"
```

**2)** I should of been clearer in the issue [here](https://github.com/bigbite/image-comparison/issues/29), so apologies for that.

The ResizableBox and WIDTH / HEIGHT settings panel should still be available when using `Align left`, `Align center` and `Align right` alignment options otherwise you cannot correctly control the block when these are set.

The reference to the [settings.layout](https://developer.wordpress.org/themes/global-settings-and-styles/settings/layout/#layout-settings) values in the issue was meant to be specifically for the values you can supply to this property in `theme.json`, so apologies for not being clear on that.

Again thanks for the work on this, if you need any assistance or would like to go over any of the points raised please feel free to reach out.

PaulREnglish committed:
refactor: set default width/height to auto

PaulREnglish committed:
fix: set container height and width to auto if align is wide or full

PaulREnglish committed:
Merge branch 'main' into fix/no-layout-alignment-override

PaulREnglish committed:
refactor: remove align check on frontend for width/height

Setting the align attribute to full or wide in the editor will automatically change the containerWidth and containerHeight to auto.  Thus we no longer need to check the align attribute on the front end to force the width and height to be auto.

jaymcp reviewed (commented):
Sorry that this has gone so long without a review. I see that if the block is align-wide or align-full, the resize handles aren't visible. However, if I set the alignment to left/centre/right, I can not resize the block, even though the handles are visible. This also happens with align-none, if I've previously selected a different alignment. I can drag the block in either direction, but it immediately snaps back to the size it became when picking the alignment.


https://github.com/user-attachments/assets/9009d00b-b6fa-4759-bcee-a5b46f57224f
